src/main/java/io/github/mzmine/modules/dataprocessing/id_localcsvsearch/DatabaseIsotopeRefinerScanBased.java

```diff
@@ -0,0 +1,220 @@
+            measuredIsotopes = ms1DefaultPattern;
+          }
+          if (measuredIsotopes == null) {
+            // no isotope pattern and no default MS1 scan. e.g., if just a feature list loaded
+            return false;
+          }
+          var score = calculateIsotopeScore(annotation, measuredIsotopes, mzTolerance, minIntensity,
+              ionIsotopePatternMap);
+          return score >= minIsotopeScore;
+        }).sorted(Comparator.comparing(CompoundDBAnnotation::getIsotopePatternScore,
+            Comparators.scoreDescending())).toList());
+
+    row.setCompoundAnnotations(remainingAnnotations);
+  }
+
+  /**
+   * @param annotation           annotation with formula and ion
+   * @param measuredIsotopes     all measured isotope signals from an IsotopePattern
+   * @param mzTolerance          matching tolerance
+   * @param minIntensity         minimum isotope intensity for prediction
+   * @param ionIsotopePatternMap map to cache ions formula to predicted patterns
+   * @return the isotope pattern score or 0 on error or if no pattern was detected
+   */
+  private static double calculateIsotopeScore(final CompoundDBAnnotation annotation,
+      final DataPoint[] measuredIsotopes, final MZTolerance mzTolerance, final double minIntensity,
+      final Map<IMolecularFormula, IsotopePattern> ionIsotopePatternMap) {
+    var adductType = annotation.getAdductType();
+    if (annotation.getFormula() == null || adductType == null) {
+      return 0;
+    }
+    // create ion formula
+    IMolecularFormula ionFormula = FormulaUtils.getIonizedFormula(annotation);
+    assert ionFormula != null;
+    // cache the ionformula to IsotopePattern to reuse isotope patterns for the same formula
+    try {
+      // TODO multiple isotope patterns for resolutions
+      final IsotopePattern predictedIsotopePattern = ionIsotopePatternMap.computeIfAbsent(
+          ionFormula,
+          key -> getCalculateIsotopePattern(mzTolerance, minIntensity, adductType, ionFormula));
+//        predictedIsotopePattern = IsotopePatternCalculator.removeDataPointsBelowIntensity(predictedIsotopePattern,
+//            minIntensity);
+      var predictedIsotopes = ScanUtils.extractDataPoints(predictedIsotopePattern);
+
+      var similarity = SpectralSimilarityFunction.compositeCosine.getSimilarity(Weights.SQRT, 0,
+          HandleUnmatchedSignalOptions.KEEP_ALL_AND_MATCH_TO_ZERO, mzTolerance, 0,
+          predictedIsotopes, measuredIsotopes);
+      // also match with library as ground truth to give more weight to predicted signals
+      var similarityLibrary = SpectralSimilarityFunction.compositeCosine.getSimilarity(Weights.SQRT,
+          0, HandleUnmatchedSignalOptions.KEEP_LIBRARY_SIGNALS, mzTolerance, 0, predictedIsotopes,
+          measuredIsotopes);
+
+      if (similarity != null && similarityLibrary != null) {
+        var score = (float) (similarity.getScore() + similarityLibrary.getScore() * 2) / 3f;
```

Comment:
why do you do this? Seems a bit random to me, wouldn't the option of choosing between one of the two scores be better?

Reply:
We can remove this later and finalize once you have the isotope patterns for different resolutions

Reply:
Was my idea on how to calculate a combined score

Reply:
ok 👍